### PR TITLE
Make Styleable.style(...) more general by using reflection

### PR DESCRIPTION
**Pull request by stubborn-me**
Rewrote Styleable.style(...) to use reflection to set the style
property instead of checking if it's one of the known classes that
contain the style property.

Now it should also work with unknown classes that also implement the
Styleable interface together with a setStyle(...) method.

Also removed TableColumnBase<*, *>.style(...) as it was redundant.

### CODE REVIEW

**Comment by edvin** [comment]
This is not a good idea, as it's going to break in Java 9. What we could do, is to continue checking for the known instances, then do reflection in a try/catch block where we only log the exception if it fails. If you change your PR to use this approach we should be good :)

**Comment by stubborn-me** [comment]
Can you point me to a link explaining why would this break in Java 9? I'm not up-to-date with Java 9 ...

**Comment by edvin** [comment]
Basically, for you to be able to reflect on a class, that class must be opened up by the library that provides it (or the jdk). This is a good fallback, as long as it runs in a try catch, but we should avoid it when possible. Would you be willing to adapt your PR? I'm hoping to do a release today, we could get this in if you have time.

**Comment by stubborn-me** [comment]
I couldn't find any information online about this limitation you just stated.

The only new limitations I found for Java 9 are about internal APIs and non-public fields, but none of those concern our case here.

Also it doesn't make much sense to limit reflection only to your own classes, reflection is almost always needed for analyzing unknown classes.

Don't know if I'm missing something major, so if you don't have time to find more details online for me, it might be easier for you to just update the code on your own, it's only about 10 lines or so...

**Comment by edvin** [comment]
In Java 9, if you don't allow reflection on your app classes, this would yield an error while the current approach would work. The current approach is also faster :) I don't have time to dig up the details right now, but you can still get the functionality you need with a simple change to your PR :)

**Comment by stubborn-me** [comment]
I hope this is what it should be.

I created a small Java 9 test program that demonstrates what I think might be the problem, I'll attach it zipped below.

Btw I didn't re-add the checking for known classes, because the reflection will pass for them anyways in Java 9, since they're all public classes.

[jdk9_reflection_test.zip](https://github.com/edvin/tornadofx/files/1421227/jdk9_reflection_test.zip)

**Comment by stubborn-me** [comment]
To spare you from opening the zip file, here are it's contents:

```
package packageB;

// this class is not visible outside packageB
class InternalClass {
	public void setStyle(String style)
	{
		System.out.println("Executing InternalClass.setStyle(...)");
	}
}

public class PublicClass
{
	public static Object createInternalObject()
	{
		return new InternalClass();
	}
}
```

```
package packageA;

import java.lang.reflect.Method;
import javafx.embed.swing.JFXPanel;
import javafx.scene.control.Tooltip;

public class Main
{
	public static void trySetter(Object obj)
	{
		try
		{
			for (Method method : obj.getClass().getMethods())
			{
				if (method.getName().equals("setStyle"))
				{
					System.out.println();
					//if (!method.isAccessible())
					if (method.canAccess(obj)) // canAccess(...) is introduced in Java 9
					{
						//System.out.println("Setting " + obj.getClass().getName() + ".setStyle(...) to accessible");
						//method.setAccessible(true);
						method.invoke(obj, "-fx-font-size: 20");
						System.out.println(obj.getClass().getName() + ".setStyle(...) was successfully invoked");
					}
					else
						System.out.println(obj.getClass().getName() + ".setStyle(...) could not be invoked due to insufficient priviledge when called from " + Main.class.getName() + ".trySetter(...)");
					return;
				}
			}
			System.out.println("Could not find an existing " + obj.getClass().getName() + ".setStyle(...) method\n");
		}
		catch (Exception e)
		{
			//System.err.println(e);
			e.printStackTrace();
		}
	}

	public static void main(String[] args)
	{
		new JFXPanel();

		trySetter(new Tooltip("some tooltip")); // invoking setStyle(...) should succeed
		trySetter(packageB.PublicClass.createInternalObject()); // invoke setStyle(...) should fail
		System.exit(0);
	}
}
```

```
Generated output:

javafx.scene.control.Tooltip.setStyle(...) was successfully invoked

packageB.InternalClass.setStyle(...) could not be invoked due to insufficient priviledge when called from packageA.Main.trySetter(...)
```

File: src/main/java/tornadofx/CSS.kt [hunk 1]
```diff
@@ -1042,30 +1042,27 @@ class InlineCss : PropertyHolder(), Rendered {
...
-
-    fun setter(value: String) = when (this) {
-        is Node -> style = value
-        is MenuItem -> style = value
-        is PopupControl -> style = value
-        is Tab -> style = value
-        is TableColumnBase<*, *> -> this.style = value
-        else -> throw IllegalArgumentException("Don't know how to set style for Styleable subclass ${this@style.javaClass}")
-    }
+    val setStyleMethod = this.javaClass.methods.firstOrNull { method -> method.name == "setStyle" }
```
Comment (edvin) [inline src/main/java/tornadofx/CSS.kt]: We should probably get the method with correct name and correct parameter count/type.

**Comment by stubborn-me** [comment]
Correct parameter count & type can be added, but I don't see why the previously supported nodes have to be re-added, I made sure that the reflection works on those as well in Java 9.

In the generated output in the test program you can see that Tooltip.setStyle(...) works without warnings in Java 9, and there are no plans for that to change in Java 10 or after.

**Comment by edvin** [comment]
The old way is faster since it doesn't require reflection, that's why I lean to keep testing for them specifically. On the other hand, I don't think the speed difference will be noticeable, so if you add a test to make sure we still set styles correctly for the types we previously supported explicitly, we should be good.

**Comment by stubborn-me** [comment]
I hope it's good now.

File: src/main/java/tornadofx/CSS.kt [hunk 2]
```diff
@@ -1042,30 +1042,27 @@ class InlineCss : PropertyHolder(), Rendered {
...
-        setter(block.render().trim())
-}
+    setStyleMethod ?: throw IllegalArgumentException("Don't know how to set style for Styleable subclass ${this@style.javaClass}")
 
-fun TableColumnBase<*, *>.style(append: Boolean = false, op: InlineCss.() -> Unit) {
     val block = InlineCss().apply(op)
+    val newStyle = when (append && style.isNotBlank()) {
+        true -> (style + block.render())
+        else -> block.render().trim()
+    }
```
[the remark this answers is not in the document]
Comment (edvin) [reply src/main/java/tornadofx/CSS.kt]: I'll just change it after merge :)

**Pull request merged**